File: client-android/app/src/main/java/stasis/client_android/persistence/config/ConfigRepository.kt
package stasis.client_android.persistence.config

import android.content.Context
import android.content.SharedPreferences
import okio.ByteString
import stasis.client_android.lib.encryption.Aes
import stasis.client_android.lib.encryption.secrets.Secret
import stasis.client_android.lib.model.server.devices.DeviceBootstrapParameters
import stasis.client_android.serialization.ByteStrings.decodeFromBase64
import stasis.client_android.serialization.ByteStrings.encodeAsBase64
import androidx.core.content.edit

class ConfigRepository(
    private val preferences: SharedPreferences,
) {
    val available: Boolean
        get() = listOfNotNull(
            preferences.getAuthenticationConfig(),
            preferences.getServerApiConfig(),
            preferences.getServerCoreConfig()
        ).isNotEmpty()

    fun bootstrap(params: DeviceBootstrapParameters) {
        preferences
            .edit(commit = true) {
                putString(Keys.Authentication.TokenEndpoint, params.authentication.tokenEndpoint)
                putString(Keys.Authentication.ClientId, params.authentication.clientId)
                putString(Keys.Authentication.ClientSecret, params.authentication.clientSecret)
                putString(Keys.Authentication.ScopeApi, params.authentication.scopes.api)
                putString(Keys.Authentication.ScopeCore, params.authentication.scopes.core)
                putString(Keys.ServerApi.Url, params.serverApi.url)
                putString(Keys.ServerApi.User, params.serverApi.user)
                putString(Keys.ServerApi.UserSalt, params.serverApi.userSalt)
                putString(Keys.ServerApi.Device, params.serverApi.device)
                putString(Keys.ServerCore.Address, params.serverCore.address)
                putString(Keys.ServerCore.NodeId, params.serverCore.nodeId)
                putInt(Keys.Secrets.DerivationEncryptionSecretSize, params.secrets.derivation.encryption.secretSize)
                putInt(Keys.Secrets.DerivationEncryptionIterations, params.secrets.derivation.encryption.iterations)
                putString(Keys.Secrets.DerivationEncryptionSaltPrefix, params.secrets.derivation.encryption.saltPrefix)
                putBoolean(
                    Keys.Secrets.DerivationAuthenticationEnabled,
                    params.secrets.derivation.authentication.enabled
                )
                putInt(
                    Keys.Secrets.DerivationAuthenticationSecretSize,
                    params.secrets.derivation.authentication.secretSize
                )
                putInt(
                    Keys.Secrets.DerivationAuthenticationIterations,
                    params.secrets.derivation.authentication.iterations
                )
                putString(
                    Keys.Secrets.DerivationAuthenticationSaltPrefix,
                    params.secrets.derivation.authentication.saltPrefix
                )
                putInt(Keys.Secrets.EncryptionFileKeySize, params.secrets.encryption.file.keySize)
                putInt(Keys.Secrets.EncryptionMetadataKeySize, params.secrets.encryption.metadata.keySize)
                putInt(Keys.Secrets.EncryptionDeviceSecretKeySize, params.secrets.encryption.deviceSecret.keySize)
            }
    }

    fun reset() {
        preferences
            .edit(commit = true) {
                remove(Keys.Authentication.TokenEndpoint)
                    .remove(Keys.Authentication.ClientId)
                    .remove(Keys.Authentication.ClientSecret)
                    .remove(Keys.Authentication.ScopeApi)
                    .remove(Keys.Authentication.ScopeCore)
                    .remove(Keys.ServerApi.Url)
                    .remove(Keys.ServerApi.User)
                    .remove(Keys.ServerApi.UserSalt)
                    .remove(Keys.ServerApi.Device)
                    .remove(Keys.ServerCore.Address)
                    .remove(Keys.ServerCore.NodeId)
                    .remove(Keys.Secrets.DerivationEncryptionSecretSize)
                    .remove(Keys.Secrets.DerivationEncryptionIterations)
                    .remove(Keys.Secrets.DerivationEncryptionSaltPrefix)
                    .remove(Keys.Secrets.DerivationAuthenticationEnabled)
                    .remove(Keys.Secrets.DerivationAuthenticationSecretSize)
                    .remove(Keys.Secrets.DerivationAuthenticationIterations)
                    .remove(Keys.Secrets.DerivationAuthenticationSaltPrefix)
                    .remove(Keys.Secrets.EncryptionFileKeySize)
                    .remove(Keys.Secrets.EncryptionMetadataKeySize)
                    .remove(Keys.Secrets.EncryptionDeviceSecretKeySize)
                    .remove(Keys.Secrets.EncryptedDeviceSecret)
                    .remove(Keys.General.IsFirstRun)
                    .remove(Keys.General.SavedUsername)
                    .remove(Keys.General.LastProcessedCommand)
                    .remove(Keys.Analytics.EntryCache)
            }
    }

    companion object {
        object Keys {
            object Authentication {
                const val TokenEndpoint: String = "authentication_token_endpoint"
                const val ClientId: String = "authentication_client_id"
                const val ClientSecret: String = "authentication_client_secret"
                const val ScopeApi: String = "authentication_scope_api"
                const val ScopeCore: String = "authentication_scope_core"

                val All: List<String> = listOf(
                    TokenEndpoint,
                    ClientId,
                    ClientSecret,
                    ScopeApi,
                    ScopeCore
                )
            }

            object ServerApi {
                const val Url: String = "server_api_url"
                const val User: String = "server_api_user"
                const val UserSalt: String = "server_api_user_salt"
                const val Device: String = "server_api_device"

                val All: List<String> = listOf(
                    Url,
                    User,
                    UserSalt,
                    Device
                )
            }

            object ServerCore {
                const val Address: String = "server_core_address"
                const val NodeId: String = "server_core_node_id"

                val All: List<String> = listOf(
                    Address,
                    NodeId
                )
            }

            object Secrets {
                const val DerivationEncryptionSecretSize: String =
                    "secrets_derivation_encryption_secrets_size"

                const val DerivationEncryptionIterations: String =
                    "secrets_derivation_encryption_iterations"

                const val DerivationEncryptionSaltPrefix: String =
                    "secrets_derivation_encryption_salt_prefix"

                const val DerivationAuthenticationEnabled: String =
                    "secrets_derivation_authentication_enabled"

                const val DerivationAuthenticationSecretSize: String =
                    "secrets_derivation_authentication_secrets_size"

                const val DerivationAuthenticationIterations: String =
                    "secrets_derivation_authentication_iterations"

                const val DerivationAuthenticationSaltPrefix: String =
                    "secrets_derivation_authentication_salt_prefix"

                const val EncryptionFileKeySize: String =
                    "secrets_encryption_file_key_size"

                const val EncryptionMetadataKeySize: String =
                    "secrets_encryption_metadata_key_size"

                const val EncryptionDeviceSecretKeySize: String =
                    "secrets_encryption_device_secret_key_size"

                const val EncryptedDeviceSecret: String =
                    "secrets_encrypted_device_secret"
            }

            object General {
                const val IsFirstRun: String = "general_is_first_run"
                const val SavedUsername: String = "general_saved_username"
                const val LastProcessedCommand: String = "last_processed_command"
            }

            object Analytics {
                const val EntryCache: String = "analytics_entry_cache"
            }
        }

        object Defaults {
            object Secrets {
                object Derivation {
                    object Encryption {
                        const val SecretSize: Int = 32
                        const val Iterations: Int = 150000
                        const val SaltPrefix: String = "changeme"
                    }

                    object Authentication {
                        const val Enabled: Boolean = true
                        const val SecretSize: Int = 16
                        const val Iterations: Int = 150000
                        const val SaltPrefix: String = "changeme"
                    }
                }

                object Encryption {
                    object File {
                        const val KeySize: Int = 16
                    }

                    object Metadata {
                        const val KeySize: Int = 16
                    }

                    object DeviceSecret {
                        const val KeySize: Int = 16
                    }
                }
            }

            object General {
                const val IsFirstRun: Boolean = true
            }
        }

        const val PreferencesFileName: String =
            "stasis.client_android.persistence.config"

        operator fun invoke(context: Context): ConfigRepository {
            return ConfigRepository(
                preferences = getPreferences(context)
            )
        }

        fun SharedPreferences.getAuthenticationConfig(): Config.Authentication? {
            val params = Keys.Authentication.All.mapNotNull { key -> this.getString(key, null) }

            val expectedParams = Keys.Authentication.All.size

            return when (params.size) {
                0 -> null
                expectedParams -> Config.Authentication(
                    tokenEndpoint = params[0],
                    clientId = params[1],
                    clientSecret = params[2],
                    scopeApi = params[3],
                    scopeCore = params[4]
                )

                else -> throw IllegalStateException(
                    "Expected [$expectedParams] authentication parameters but [${params.size}] found"
                )
            }
        }

        fun SharedPreferences.getServerApiConfig(): Config.ServerApi? {
            val params = Keys.ServerApi.All.mapNotNull { key -> this.getString(key, null) }

            val expectedParams = Keys.ServerApi.All.size

            return when (params.size) {
                0 -> null
                expectedParams -> Config.ServerApi(
                    url = params[0],
                    user = params[1],
                    userSalt = params[2],
                    device = params[3]
                )

                else -> throw IllegalStateException(
                    "Expected [$expectedParams] server API parameters but [${params.size}] found"
                )
            }
        }

        fun SharedPreferences.getServerCoreConfig(): Config.ServerCore? {
            val params = Keys.ServerCore.All.mapNotNull { key -> this.getString(key, null) }

            val expectedParams = Keys.ServerCore.All.size

            return when (params.size) {
                0 -> null
                expectedParams -> Config.ServerCore(
                    address = params[0],
                    nodeId = params[1]
                )

                else -> throw IllegalStateException(
                    "Expected [$expectedParams] server core parameters but [${params.size}] found"
                )
            }
        }

        fun SharedPreferences.getSecretsConfig(): Secret.Config = Secret.Config(
            derivation = Secret.Config.DerivationConfig(
                encryption = Secret.EncryptionKeyDerivationConfig(
                    secretSize = this.getInt(
                        Keys.Secrets.DerivationEncryptionSecretSize,
                        Defaults.Secrets.Derivation.Encryption.SecretSize
                    ),
                    iterations = this.getInt(
                        Keys.Secrets.DerivationEncryptionIterations,
                        Defaults.Secrets.Derivation.Encryption.Iterations
                    ),
                    saltPrefix = this.getString(
                        Keys.Secrets.DerivationEncryptionSaltPrefix,
                        Defaults.Secrets.Derivation.Encryption.SaltPrefix
                    ) ?: Defaults.Secrets.Derivation.Encryption.SaltPrefix
                ),
                authentication = Secret.AuthenticationKeyDerivationConfig(
                    enabled = this.getBoolean(
                        Keys.Secrets.DerivationAuthenticationEnabled,
                        Defaults.Secrets.Derivation.Authentication.Enabled
                    ),
                    secretSize = this.getInt(
                        Keys.Secrets.DerivationAuthenticationSecretSize,
                        Defaults.Secrets.Derivation.Authentication.SecretSize
                    ),
                    iterations = this.getInt(
                        Keys.Secrets.DerivationAuthenticationIterations,
                        Defaults.Secrets.Derivation.Authentication.Iterations
                    ),
                    saltPrefix = this.getString(
                        Keys.Secrets.DerivationAuthenticationSaltPrefix,
                        Defaults.Secrets.Derivation.Authentication.SaltPrefix
                    ) ?: Defaults.Secrets.Derivation.Authentication.SaltPrefix
                )
            ),
            encryption = Secret.Config.EncryptionConfig(
                file = Secret.EncryptionSecretConfig(
                    keySize = this.getInt(
                        Keys.Secrets.EncryptionFileKeySize,
                        Defaults.Secrets.Encryption.File.KeySize
                    ),
                    ivSize = Aes.IvSize
                ),
                metadata = Secret.EncryptionSecretConfig(
                    keySize = this.getInt(
                        Keys.Secrets.EncryptionMetadataKeySize,
                        Defaults.Secrets.Encryption.Metadata.KeySize
                    ),
                    ivSize = Aes.IvSize
                ),
                deviceSecret = Secret.EncryptionSecretConfig(
                    keySize = this.getInt(
                        Keys.Secrets.EncryptionDeviceSecretKeySize,
                        Defaults.Secrets.Encryption.DeviceSecret.KeySize
                    ),
                    ivSize = Aes.IvSize
                )
            )
        )

        fun SharedPreferences.putEncryptedDeviceSecret(secret: ByteString) {
            this.edit(commit = true) {
                putString(Keys.Secrets.EncryptedDeviceSecret, secret.encodeAsBase64())
            }
        }

        fun SharedPreferences.getEncryptedDeviceSecret(): ByteString {
            val encryptedDeviceSecret = this
                .getString(Keys.Secrets.EncryptedDeviceSecret, null)
                ?.decodeFromBase64()

            require(encryptedDeviceSecret != null) {
                "Expected device secret with key [${Keys.Secrets.EncryptedDeviceSecret}] but none was found"
            }

            return encryptedDeviceSecret
        }

        fun SharedPreferences.isFirstRun(): Boolean =
            this.getBoolean(Keys.General.IsFirstRun, Defaults.General.IsFirstRun)

        fun SharedPreferences.firstRunComplete() {
            this.edit(commit = true) {
                putBoolean(Keys.General.IsFirstRun, false)
            }
        }

        fun SharedPreferences.savedUsername(): String? =
            this.getString(Keys.General.SavedUsername, null)

        fun SharedPreferences.saveUsername(username: String?) {
            this.edit(commit = true) {
                putString(Keys.General.SavedUsername, username)
            }
        }

        fun SharedPreferences.savedLastProcessedCommand(): Long =
            this.getLong(Keys.General.LastProcessedCommand, 0L)

        fun SharedPreferences.saveLastProcessedCommand(sequenceId: Long) {
            this.edit(commit = true) {
                putLong(Keys.General.LastProcessedCommand, sequenceId)
            }
        }

        fun SharedPreferences.getAnalyticsCachedEntry(): String? =
            this.getString(Keys.Analytics.EntryCache, null)

        fun SharedPreferences.putAnalyticsCachedEntry(entry: String) {
            this.edit(commit = true) {
                putString(Keys.Analytics.EntryCache, entry)
            }
        }

        fun getPreferences(context: Context): SharedPreferences =
            context.getSharedPreferences(PreferencesFileName, Context.MODE_PRIVATE)
    }
}

File: client-ios/app/StasisClient/Persistence/Config/ConfigRepository.swift
import Foundation
import StasisClientLib

public final class ConfigRepository: @unchecked Sendable {
    public static let suiteName: String = "group.stasis.client.ios"

    public enum Keys {
        public enum Authentication {
            public static let tokenEndpoint = "authentication_token_endpoint"
            public static let clientId = "authentication_client_id"
            public static let clientSecret = "authentication_client_secret"
            public static let scopeApi = "authentication_scope_api"
            public static let scopeCore = "authentication_scope_core"
            public static let all: [String] = [tokenEndpoint, clientId, clientSecret, scopeApi, scopeCore]
        }

        public enum ServerApi {
            public static let url = "server_api_url"
            public static let user = "server_api_user"
            public static let userSalt = "server_api_user_salt"
            public static let device = "server_api_device"
            public static let all: [String] = [url, user, userSalt, device]
        }

        public enum ServerCore {
            public static let address = "server_core_address"
            public static let nodeId = "server_core_node_id"
            public static let all: [String] = [address, nodeId]
        }

        public enum Secrets {
            public static let derivationEncryptionSecretSize = "secrets_derivation_encryption_secrets_size"
            public static let derivationEncryptionIterations = "secrets_derivation_encryption_iterations"
            public static let derivationEncryptionSaltPrefix = "secrets_derivation_encryption_salt_prefix"
            public static let derivationAuthenticationEnabled = "secrets_derivation_authentication_enabled"
            public static let derivationAuthenticationSecretSize = "secrets_derivation_authentication_secrets_size"
            public static let derivationAuthenticationIterations = "secrets_derivation_authentication_iterations"
            public static let derivationAuthenticationSaltPrefix = "secrets_derivation_authentication_salt_prefix"
            public static let encryptionFileKeySize = "secrets_encryption_file_key_size"
            public static let encryptionMetadataKeySize = "secrets_encryption_metadata_key_size"
            public static let encryptionDeviceSecretKeySize = "secrets_encryption_device_secret_key_size"
            public static let encryptedDeviceSecret = "secrets_encrypted_device_secret"
        }

        public enum General {
            public static let isFirstRun = "general_is_first_run"
            public static let savedUsername = "general_saved_username"
            public static let lastProcessedCommand = "last_processed_command"
        }

        public enum Analytics {
            public static let entryCache = "analytics_entry_cache"
            public static let pendingCache = "analytics_pending_cache"
        }
    }

    public enum Defaults {
        public enum Secrets {
            public enum Derivation {
                public enum Encryption {
                    public static let secretSize: Int = 32
                    public static let iterations: Int = 150_000
                    public static let saltPrefix: String = "changeme"
                }

                public enum Authentication {
                    public static let enabled: Bool = true
                    public static let secretSize: Int = 16
                    public static let iterations: Int = 150_000
                    public static let saltPrefix: String = "changeme"
                }
            }

            public enum Encryption {
                public enum File {
                    public static let keySize: Int = 16
                }

                public enum Metadata {
                    public static let keySize: Int = 16
                }

                public enum DeviceSecret {
                    public static let keySize: Int = 16
                }
            }
        }

        public enum General {
            public static let isFirstRun: Bool = true
        }
    }

    public enum RepositoryError: Error, Equatable, LocalizedError {
        case malformedConfig(String)
        case missingDeviceSecret

        public var errorDescription: String? {
            switch self {
            case .malformedConfig(let reason):
                "Malformed configuration: \(reason)"
            case .missingDeviceSecret:
                "No device secret is available"
            }
        }
    }

    private let preferences: UserDefaults

    public init(preferences: UserDefaults) {
        self.preferences = preferences
    }

    public convenience init?(suiteName: String = ConfigRepository.suiteName) {
        guard let defaults = UserDefaults(suiteName: suiteName) else { return nil }
        self.init(preferences: defaults)
    }

    public func available() throws -> Bool {
        let configs: [Any?] = try [
            preferences.authenticationConfig(),
            preferences.serverApiConfig(),
            preferences.serverCoreConfig()
        ]
        return configs.contains { $0 != nil }
    }

    public func bootstrap(params: DeviceBootstrapParameters) {
        preferences.set(params.authentication.tokenEndpoint, forKey: Keys.Authentication.tokenEndpoint)
        preferences.set(params.authentication.clientId, forKey: Keys.Authentication.clientId)
        preferences.set(params.authentication.clientSecret, forKey: Keys.Authentication.clientSecret)
        preferences.set(params.authentication.scopes.api, forKey: Keys.Authentication.scopeApi)
        preferences.set(params.authentication.scopes.core, forKey: Keys.Authentication.scopeCore)
        preferences.set(params.serverApi.url, forKey: Keys.ServerApi.url)
        preferences.set(params.serverApi.user, forKey: Keys.ServerApi.user)
        preferences.set(params.serverApi.userSalt, forKey: Keys.ServerApi.userSalt)
        preferences.set(params.serverApi.device, forKey: Keys.ServerApi.device)
        preferences.set(params.serverCore.address, forKey: Keys.ServerCore.address)
        preferences.set(params.serverCore.nodeId, forKey: Keys.ServerCore.nodeId)
        preferences.set(
            params.secrets.derivation.encryption.secretSize,
            forKey: Keys.Secrets.derivationEncryptionSecretSize
        )
        preferences.set(
            params.secrets.derivation.encryption.iterations,
            forKey: Keys.Secrets.derivationEncryptionIterations
        )
        preferences.set(
            params.secrets.derivation.encryption.saltPrefix,
            forKey: Keys.Secrets.derivationEncryptionSaltPrefix
        )
        preferences.set(
            params.secrets.derivation.authentication.enabled,
            forKey: Keys.Secrets.derivationAuthenticationEnabled
        )
        preferences.set(
            params.secrets.derivation.authentication.secretSize,
            forKey: Keys.Secrets.derivationAuthenticationSecretSize
        )
        preferences.set(
            params.secrets.derivation.authentication.iterations,
            forKey: Keys.Secrets.derivationAuthenticationIterations
        )
        preferences.set(
            params.secrets.derivation.authentication.saltPrefix,
            forKey: Keys.Secrets.derivationAuthenticationSaltPrefix
        )
        preferences.set(params.secrets.encryption.file.keySize, forKey: Keys.Secrets.encryptionFileKeySize)
        preferences.set(params.secrets.encryption.metadata.keySize, forKey: Keys.Secrets.encryptionMetadataKeySize)
        preferences.set(
            params.secrets.encryption.deviceSecret.keySize,
            forKey: Keys.Secrets.encryptionDeviceSecretKeySize
        )
    }

    public func reset() {
        let keys: [String] = Keys.Authentication.all
            + Keys.ServerApi.all
            + Keys.ServerCore.all
            + [
                Keys.Secrets.derivationEncryptionSecretSize,
                Keys.Secrets.derivationEncryptionIterations,
                Keys.Secrets.derivationEncryptionSaltPrefix,
                Keys.Secrets.derivationAuthenticationEnabled,
                Keys.Secrets.derivationAuthenticationSecretSize,
                Keys.Secrets.derivationAuthenticationIterations,
                Keys.Secrets.derivationAuthenticationSaltPrefix,
                Keys.Secrets.encryptionFileKeySize,
                Keys.Secrets.encryptionMetadataKeySize,
                Keys.Secrets.encryptionDeviceSecretKeySize,
                Keys.Secrets.encryptedDeviceSecret,
                Keys.General.isFirstRun,
                Keys.General.savedUsername,
                Keys.General.lastProcessedCommand,
                Keys.Analytics.entryCache,
                Keys.Analytics.pendingCache
            ]
        for key in keys { preferences.removeObject(forKey: key) }
    }

    public var preferencesStore: UserDefaults { preferences }
}

public extension UserDefaults {
    func authenticationConfig() throws -> Config.Authentication? {
        let params = ConfigRepository.Keys.Authentication.all.compactMap { string(forKey: $0) }
        switch params.count {
        case 0:
            return nil
        case ConfigRepository.Keys.Authentication.all.count:
            return Config.Authentication(
                tokenEndpoint: params[0],
                clientId: params[1],
                clientSecret: params[2],
                scopeApi: params[3],
                scopeCore: params[4]
            )
        default:
            throw ConfigRepository.RepositoryError.malformedConfig(
                "Expected [\(ConfigRepository.Keys.Authentication.all.count)] authentication parameters" +
                    " but [\(params.count)] found"
            )
        }
    }

    func serverApiConfig() throws -> Config.ServerApi? {
        let params = ConfigRepository.Keys.ServerApi.all.compactMap { string(forKey: $0) }
        switch params.count {
        case 0:
            return nil
        case ConfigRepository.Keys.ServerApi.all.count:
            return Config.ServerApi(
                url: params[0],
                user: params[1],
                userSalt: params[2],
                device: params[3]
            )
        default:
            throw ConfigRepository.RepositoryError.malformedConfig(
                "Expected [\(ConfigRepository.Keys.ServerApi.all.count)] server API parameters" +
                    " but [\(params.count)] found"
            )
        }
    }

    func serverCoreConfig() throws -> Config.ServerCore? {
        let params = ConfigRepository.Keys.ServerCore.all.compactMap { string(forKey: $0) }
        switch params.count {
        case 0:
            return nil
        case ConfigRepository.Keys.ServerCore.all.count:
            return Config.ServerCore(address: params[0], nodeId: params[1])
        default:
            throw ConfigRepository.RepositoryError.malformedConfig(
                "Expected [\(ConfigRepository.Keys.ServerCore.all.count)] server core parameters" +
                    " but [\(params.count)] found"
            )
        }
    }

    func secretsConfig() throws -> SecretConfig {
        try SecretConfig(
            derivation: .init(
                encryption: derivationEncryptionConfig(),
                authentication: derivationAuthenticationConfig()
            ),
            encryption: .init(
                file: encryptionConfig(
                    keySizeKey: ConfigRepository.Keys.Secrets.encryptionFileKeySize,
                    defaultKeySize: ConfigRepository.Defaults.Secrets.Encryption.File.keySize
                ),
                metadata: encryptionConfig(
                    keySizeKey: ConfigRepository.Keys.Secrets.encryptionMetadataKeySize,
                    defaultKeySize: ConfigRepository.Defaults.Secrets.Encryption.Metadata.keySize
                ),
                deviceSecret: encryptionConfig(
                    keySizeKey: ConfigRepository.Keys.Secrets.encryptionDeviceSecretKeySize,
                    defaultKeySize: ConfigRepository.Defaults.Secrets.Encryption.DeviceSecret.keySize
                )
            )
        )
    }

    private func derivationEncryptionConfig() throws -> EncryptionKeyDerivationConfig {
        try EncryptionKeyDerivationConfig(
            secretSize: integerOrDefault(
                forKey: ConfigRepository.Keys.Secrets.derivationEncryptionSecretSize,
                default: ConfigRepository.Defaults.Secrets.Derivation.Encryption.secretSize
            ),
            iterations: integerOrDefault(
                forKey: ConfigRepository.Keys.Secrets.derivationEncryptionIterations,
                default: ConfigRepository.Defaults.Secrets.Derivation.Encryption.iterations
            ),
            saltPrefix: string(forKey: ConfigRepository.Keys.Secrets.derivationEncryptionSaltPrefix)
                ?? ConfigRepository.Defaults.Secrets.Derivation.Encryption.saltPrefix
        )
    }

    private func derivationAuthenticationConfig() throws -> AuthenticationKeyDerivationConfig {
        try AuthenticationKeyDerivationConfig(
            enabled: value(forKey: ConfigRepository.Keys.Secrets.derivationAuthenticationEnabled) as? Bool
                ?? ConfigRepository.Defaults.Secrets.Derivation.Authentication.enabled,
            secretSize: integerOrDefault(
                forKey: ConfigRepository.Keys.Secrets.derivationAuthenticationSecretSize,
                default: ConfigRepository.Defaults.Secrets.Derivation.Authentication.secretSize
            ),
            iterations: integerOrDefault(
                forKey: ConfigRepository.Keys.Secrets.derivationAuthenticationIterations,
                default: ConfigRepository.Defaults.Secrets.Derivation.Authentication.iterations
            ),
            saltPrefix: string(forKey: ConfigRepository.Keys.Secrets.derivationAuthenticationSaltPrefix)
                ?? ConfigRepository.Defaults.Secrets.Derivation.Authentication.saltPrefix
        )
    }

    private func encryptionConfig(keySizeKey: String, defaultKeySize: Int) throws -> EncryptionSecretConfig {
        try EncryptionSecretConfig(
            keySize: integerOrDefault(forKey: keySizeKey, default: defaultKeySize),
            ivSize: Aes.ivSize
        )
    }

    func putEncryptedDeviceSecret(_ secret: Data) {
        set(secret.base64EncodedString(), forKey: ConfigRepository.Keys.Secrets.encryptedDeviceSecret)
    }

    func encryptedDeviceSecret() throws -> Data {
        guard let raw = string(forKey: ConfigRepository.Keys.Secrets.encryptedDeviceSecret),
              let decoded = Data(base64Encoded: raw) else {
            throw ConfigRepository.RepositoryError.missingDeviceSecret
        }
        return decoded
    }

    func isFirstRun() -> Bool {
        value(forKey: ConfigRepository.Keys.General.isFirstRun) as? Bool
            ?? ConfigRepository.Defaults.General.isFirstRun
    }

    func firstRunComplete() {
        set(false, forKey: ConfigRepository.Keys.General.isFirstRun)
    }

    func savedUsername() -> String? {
        string(forKey: ConfigRepository.Keys.General.savedUsername)
    }

    func saveUsername(_ username: String?) {
        if let username {
            set(username, forKey: ConfigRepository.Keys.General.savedUsername)
        } else {
            removeObject(forKey: ConfigRepository.Keys.General.savedUsername)
        }
    }

    func savedLastProcessedCommand() -> Int64 {
        (value(forKey: ConfigRepository.Keys.General.lastProcessedCommand) as? NSNumber)?.int64Value ?? 0
    }

    func saveLastProcessedCommand(_ sequenceId: Int64) {
        set(NSNumber(value: sequenceId), forKey: ConfigRepository.Keys.General.lastProcessedCommand)
    }

    func analyticsCachedEntry() -> String? {
        string(forKey: ConfigRepository.Keys.Analytics.entryCache)
    }

    func putAnalyticsCachedEntry(_ entry: String) {
        set(entry, forKey: ConfigRepository.Keys.Analytics.entryCache)
    }

    func analyticsPendingEntries() -> String? {
        string(forKey: ConfigRepository.Keys.Analytics.pendingCache)
    }

    func putAnalyticsPendingEntries(_ entries: String) {
        set(entries, forKey: ConfigRepository.Keys.Analytics.pendingCache)
    }

    private func integerOrDefault(forKey key: String, default fallback: Int) -> Int {
        guard let value = value(forKey: key) as? NSNumber else { return fallback }
        return value.intValue
    }
}
